### PR TITLE
Clarify timeout and terminate_after docs

### LINKED ISSUE #47716: Search request timeout option is confusing
The documentation around the `timeout` option of the search request is confusing for users. Most of the time it is understood as a way to cancel any search request that takes longer than the configured threshold and users don't expect to get partial results in the response. In fact they don't even expect to get a response. While the documentation could be improved I also wonder if we should not renamed the option or simply remove it. We have a new way to cancel search requests efficiently from the client in 7.4 (by closing the underlying http channel) and we expose this in our official Java client so we don't need this option on the server if the goal is to trash any requests that take more than a configured threshold. One option I was thinking regarding the renaming was to handle time unit in the `terminate_after` option, today users can set the `terminate_after` option to control the max number of documents that should be collected per shard but we could also accept a time as in `terminate_after=10ms`. IMO this would be less confusing and would leave any real timeout option to the client side where it belongs. I am also curious to hear if this feature is really used to get partial results since that's the only reason why it should be preferred over a timeout on the client side.

### PR DESCRIPTION
**Pull request by wylieconlon**
It looks like @jimczi started an effort to rename the `timeout` setting to `shard_timeout`, which included a docs update. This PR takes just the docs updates and re-applies them, with a few tweaks due to changes since that PR was originally opened.

Reference PR: https://github.com/elastic/elasticsearch/pull/51858/

Closes https://github.com/elastic/elasticsearch/issues/47716

### CODE REVIEW

File: docs/reference/rest-api/common-parms.asciidoc
Comment (jrodewig) [inline docs/reference/rest-api/common-parms.asciidoc:1083]:
```suggestion
(Optional, integer) Maximum number of documents to collect for each shard. If a
query reaches this limit, {es} terminates the query early. {es} collects
documents before sorting.
+
IMPORTANT: When possible, {es} performs early termination automatically. We do
not recommend specifying this parameter.
```

File: docs/reference/search/search-your-data/search-your-data.asciidoc
Comment (jrodewig) [inline docs/reference/search/search-your-data/search-your-data.asciidoc:243]:
```suggestion
like to wait on each shard to complete. Each shard collects hits within the
specified time period. If collection isn't finished when the period ends, {es}
uses only the hits accumulated up to that point. The overall latency of a search
request depends on the number of shards needed for the search and the number of
concurrent shard requests.
```

**Review by jrodewig — COMMENTED** [review]
Thanks @wylieconlon. I left a few wording suggestions.

I'll let @jimczi take a look before merging this in.

File: docs/reference/rest-api/common-parms.asciidoc
Comment (jrodewig) [inline docs/reference/rest-api/common-parms.asciidoc:1085]:
```suggestion
IMPORTANT: Use with caution. {es} applies this parameter to each shard handling
the request. When possible, let {es} perform early termination automatically. We
do not recommend specifying this parameter for requests that target data streams
with backing indices across multiple data tiers.
```

Comment (jrodewig) [reply docs/reference/rest-api/common-parms.asciidoc]: Thanks for the clarification @jimczi. I pushed 0f84337e41e7cfbd58426f3c876536c8c8bcdb0a to address. Let me know if the updated copy looks okay to you.

Comment (jimczi) [reply docs/reference/rest-api/common-parms.asciidoc]: thanks

**Review by jimczi — APPROVED** [review]
LGTM

**Comment by jrodewig** [comment]
I'll merge and backport this. Thanks again @wylieconlon.

**Pull request merged**
